### PR TITLE
Safe Buffer usage

### PR DESCRIPTION
**Pull request by dcousens**
@fanatid review would be appreciated :+1: .

Non-controversial,  but, sensitive and I'd like at least 2 extra eyes since it touches the ECDSA module.

This is **not** a breaking change (we're already at a minimum of Node 4.5.0)

### CODE REVIEW

**Comment by fanatid** [comment]
I like `from`/`alloc`/`allocUnsafe`, but this is not supported by node < 4.5.0 :(
do you sure about this?

**Comment by dcousens** [comment]
@fanatid we already bumped to `4.5.0` minimum with `3.0.0` release whether we like it or not,  as we use `bs58check`@2.0.0 ... we could revert that if necessary, but doesn't change the fact that this PR, right now,  isn't a breaking change.

**Comment by dcousens** [comment]
Ref https://github.com/bitcoinjs/bitcoinjs-lib/pull/741

File: src/script_number.js [hunk 1]
```diff
@@ -16,8 +16,8 @@ function decode (buffer, maxLength, minimal) {
     var a = buffer.readUInt32LE(0)
     var b = buffer.readUInt8(4)
 
-    if (b & 0x80) return -((b & ~0x80) * 0x100000000 + a)
-    return b * 0x100000000 + a
+    if (b & 0x80) return -(((b & ~0x80) * 0x100000000) + a)
+    return (b * 0x100000000) + a
```
Comment (fanatid) [inline src/script_number.js]: extra parentheses for readable?

Comment (dcousens) [inline src/script_number.js]: `standard`@latest warns on these, haven't bumped yet in repository though

**Comment by fanatid** [comment]
Currently `engines.node` is `>=4.0.0`, but all nodes < 4.5.0 not support from/alloc/allocUnsafe ...
https://github.com/bitcoinjs/bitcoinjs-lib/blob/v3.0.0/package.json#L7

File: src/address.js [hunk 2]
```diff
@@ -18,7 +18,7 @@ function fromBase58Check (address) {
 function toBase58Check (hash, version) {
   typeforce(types.tuple(types.Hash160bit, types.UInt8), arguments)
 
-  var payload = new Buffer(21)
+  var payload = Buffer.allocUnsafe(21)
```
Comment (fanatid) [inline src/address.js]: I see two variants of this PR: `Buffer.alloc(length)` and `Buffer.allocUnsafe(length)`. Can you choose one?

Comment (dcousens) [inline src/address.js]: @fanatid I've used `Buffer.alloc` all through the tests... and `Buffer.allocUnsafe` in parts of the code where were previously using `new Buffer`; so no change.

We should probably only need need `Buffer.alloc` in cases of user input no?

**Comment by dcousens** [comment]
Wait... `bs58check` was breaking because it relied on `bs58` which had the `Buffer.from` introduced... but that has since been reverted and we only lost `0.12`... so actually this **is** a breaking change.

Blimey.
Waiting on travis to confirm.

**Comment by fanatid** [comment]
travis will be fine, because `4` there is a link to latest 4.x

**Comment by fanatid** [comment]
base-x was bumped correctly https://github.com/cryptocoinjs/base-x/blob/v3.0.0/package.json#L41

**Comment by dcousens** [comment]
@fanatid only [after a screw around](https://github.com/cryptocoinjs/base-x/issues/35#issuecomment-279305099).
Hence my confusion; and why yes,  **this is a breaking change**.

File: src/address.js (hunk 2 above)
Comment (fanatid) [inline src/address.js]: `allocUnsafe` (from docs) uses internal memory pool, maybe we should use `alloc` everywhere?

Comment (dcousens) [inline src/address.js]: I think this code is tested well enough that we can use both?

**Comment by dcousens** [comment]
Re-targeting `3.0.0` by using `safe-buffer`

**Comment by dcousens** [comment]
Rebased,  using `safe-buffer` and good to merge for a patch release

**Pull request merged**